### PR TITLE
[FLINK-15546][table-planner-blink] Fix obscure error message from Sca…

### PR DESCRIPTION
**Pull request by docete**
…larOperatorGens::generateCast

## What is the purpose of the change

ScalarOperatorGens::generateCast would produce obscure error message such as `Unsupported cast from 'ROW' to 'ROW'.` if we want cast one RowType to another.Users are unlikely to figure out what goes wrong from the above error message. Something like `Unsupported cast from 'ROW<DECIMAL(2,1), CHAR(3)>' to 'ROW<DOUBLE, VARCHAR(10)>' `will be more helpful. This PR would fix this.

## Brief change log

- 0396817 Use `LogicalType` instead of `LogicalTypeRoot` to produce clear error message.

## Verifying this change

This change added tests to verify

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

**Comment by docete** [comment]
Contact with Python Team and the python test failed because of FLINK-15929 (The parent process is killed for timeout when downloading dependencies).

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/RowTypeTest.scala [hunk 1]
```diff
@@ -129,4 +128,17 @@ class RowTypeTest extends RowTypeTestBase {
...
   }
+
+  @Test
+  def testUnsupportedCast(): Unit = {
+    expectedException.expect(classOf[ValidationException])
+    expectedException.expectMessage(
+      "Unsupported cast from Row(f0: String, f1: Boolean) to Long")
+
+    testTableApi(
+      'f5.cast(DataTypes.BIGINT()),
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/RowTypeTest.scala]: This test case can't reproduce the problem which reported by this issue. You should add a test case which cast a ROW to another ROW, e.g. 

```scala
testSqlApi("CAST(f5 AS ROW<f0 INT, f1 STRING>)", "")
```

Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/RowTypeTest.scala]: Ah... yes, this is table api, we want to fix sql api.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/ScalarOperatorGens.scala [hunk 2]
```diff
@@ -1280,8 +1280,8 @@ object ScalarOperatorGens {
           | (INTERVAL_YEAR_MONTH, BIGINT) =>
       internalExprCasting(operand, targetType)
 
-    case (from, to) =>
-      throw new CodeGenException(s"Unsupported cast from '$from' to '$to'.")
+    case (_, _) =>
+      throw new CodeGenException(s"Unsupported cast from '${operand.resultType}' to '$targetType'.")
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/ScalarOperatorGens.scala]: Nit: Add `''` to table api too.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/RowTypeTest.scala (hunk 1 above)
Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/RowTypeTest.scala]: You should keep table api and sql api both.

Comment (docete) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/RowTypeTest.scala]: OK, Added a sql api test.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/ScalarOperatorGens.scala (hunk 2 above)
Comment (docete) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/ScalarOperatorGens.scala]: modified

**Comment by docete** [comment]
@wuchong pls take another look.

**Comment by docete** [comment]
Azure fails for unstable test and fixed in https://issues.apache.org/jira/browse/FLINK-15811

**Pull request merged**